### ISSUE TITLE
Make color checker module compatible with RGB Cube LUTs

### ISSUE THREAD

**Issue by aurelienpierre** [darktable-org/darktable#1981]
There are plenty of good film-emulation LUTs out there (https://www.colorgradingcentral.com/free-luts/) used for video editors. These are RGB LUTs taking usually log-encoded input from cameras, and outputing linear RGB (ACES or REC 2020).

Making darktable colorchecker module able to read these LUTs would enable users to take advantage of the existing film-emulations databases. These are really difficult to reproduce with the color-checker tool because they need a film scanner properly neutralized/calibrated, to provide a scanner-agnostic film reference, then a properly neutralized raw file, to provide a camera-agnostic base for the LUT grading, and only then it becomes possible to transfer the film look. This is not possible without a full lab and adequate ressources, and darktable options only allow local work-arounds. Also, we don't want to fully apply the film characteristics (especially the tone curve), since its dynamic range can be more limited than digital for color emulsions, but we might want to retain the advantages of digital (shadows recovery) and the feeling of the film (color, smooth transitions).

Instead, making the colorchecker capable of reading cube LUTs would require a log transformation of the input, then a modification of the actual algo from Lab to RGB. The output would need to be converted back to Lab, although if #1841 achieves to provide a full RGB pipe, that would not be necessary anymore.

Technical details on the 3D Cube LUTs can be found here : 

- http://download.autodesk.com/us/systemdocs/pdf/lustre_color_management_user_guide.pdf
- http://juanmelara.com.au/blog/print-film-emulation-luts-for-download

**Comment by aurelienpierre** [comment]
Note that @houz has something like that working, but not it dt

**Comment by phweyland** [comment]
Looking a bit at this, colorchecker is designed for 49 patches max and is a kind of 1D LUT file (while not RGB but Lab).
3D LUT files can have 256x256x256 patches max. I've opened some with LUT_3D_SIZE 64 (64x64x64 patches). The corresponding file size is above 7Mb. That will impact seriously the preset size.

>>Note that @houz has something like that working, but not it dt
Where we can find this ?

**Comment by aurelienpierre** [comment]
> The corresponding file size is above 7Mb.

Are you sure ? The LUTs I have tested weight 970 kB max. Given that they are only text files, 7 MB seems quite heavy. Also, we don't have to store them in the database, they can be retrieved from the files like ICC profiles when needed.

**Comment by phweyland** [comment]
970Kb is for a LUT size = 33 while 7Mb is for LUT size = 66. The one I've downloaded is SLog_WB_Blue.cube but I don't remember where I've got it from.
Is it possible to look at the@houz work ?

**Comment by aurelienpierre** [comment]
>  Is it possible to look at the@houz work ?

I don't think it's commited somewhere yet.

**Comment by phweyland** [comment]
photoflow can read cube files, at least SIZE=33 (SIZE=66 doesn't end). It uses rawtherapee code.
RT seems not to read cube files, but png files instead (HaldCLUT) coming from G'MIC.

**Comment by phweyland** [comment]
What you describe in the initial post shows that usages are diverse and can be applied at different stages of the development.
Then which RGB space the LUT should be applied to ?

**Comment by aurelienpierre** [comment]
>  Then which RGB space the LUT should be applied to ?

Only in the RGB space in which the LUT has been generated. Usually, it's REC2020 or ACES P0.

**Comment by phweyland** [comment]
> Usually, it's REC2020 or ACES P0.

RT, using haldclut png files, checks and uses the lut color space. 
Cube file doesn’t contain any color space information.
The lut files can be used also for color space transformation, in fact for any transformation, including mimic a full dt style, for example.
I haven’t the answer, but I think it may be more complex than referring automatically to REC2020 or ACES P0, though they could be the default space. Depending on the rgb space used the rendering will be different, won’t it be?

**Comment by aurelienpierre** [comment]
> I haven’t the answer, but I think it may be more complex than referring automatically to REC2020 or ACES P0, though they could be the default space.

That piece of information is to be provided by the LUT maker.

 > Depending on the rgb space used the rendering will be different, won’t it be?

Indeed, if you think of an RGB space as a vector sub-space of XYZ, different RGB primaries means different base vector length, so distances are scaled differently. Given that the LUT essentially shifts the RGB coordinates, the shifts have different meanings in XYZ, depending on the RGB space used.

**Comment by phweyland** [comment]
> Usually, it's REC2020 or ACES P0.

RT, using haldclut png files, checks and uses the lut color space. 
Cube file doesn’t contain any color space information.
The lut files can be used also for color space transformation, in fact for any transformation, including mimic a full dt style, for example.
I haven’t the answer, but I think it may be more complex than referring automatically to REC2020 or ACES P0, though they could be the default space. Depending on the rgb space used the rendering will be different, won’t it be?

**Comment by phweyland** [comment]
I would to start on this topic.
I've found a module which have never been merged:  haldcut (png file) import (https://github.com/PkmX/darktable). I've tested it and it works.
Having a specific module seems to me more appropriate and simpler than derivating color checker which more an UI to manage color patches.
Starting from there, I would rename it to something like lut3d, and add the following features:
- select and open cube files
- select application colorspace (sRGB, linear RGB, Lab, XYZ, ...).
